### PR TITLE
Split up the test library into individual classes

### PR DESCRIPTION
**Pull request by FrBier**
'modtestlib.py' has been renamed to 'modtestclasses.py' and split up into 3 separate classes,
one to drive the Debug NoC, one for register accesses and one for various utilities:

NoC_Driver: class for initializing the debug test bench and sending individual flits
Reg_Access: class for exchanging information on packet level
Utility: class for composing payloads, evaluating packets and conducting tests

Code in the library file and the test file has been adjusted accordingly.

Coding style has been improved by:
- using a smaller line width
- using keyword only functions
- using dut._log.info() to log messages instead of print()
- removing CONST_ prefix constants

### CODE REVIEW

File: test/cocotb/osdtestlib/modtestclasses.py [hunk 1]
```diff
@@ -0,0 +1,637 @@
+import cocotb
+from cocotb.triggers import Timer
+from cocotb.clock import Clock
+from cocotb.result import TestFailure
+from cocotb.triggers import RisingEdge
+
+class NoC_Driver:
```
Comment (imphil) [inline test/cocotb/osdtestlib/modtestclasses.py]: Please use a camel-case name for the class (https://www.python.org/dev/peps/pep-0008/#class-names).

File: test/cocotb/osdtestlib/modtestclasses.py [hunk 2]
```diff
@@ -0,0 +1,637 @@
...
+    """
+    Class responsible for delivering individual words to a debug module and 
+    processing their reaction
+    """
+
+    @cocotb.coroutine
+    def build_flit(self, dut, data, last):
+        """function which transmits one flit between debug modules
```
Comment (imphil) [inline test/cocotb/osdtestlib/modtestclasses.py]: The method name says a flit is built, the comment says it's transmitted. What's correct?

Also, there's no need to include the "function which" part in the documentation. Simply write "Transmit one flit between modules".

File: test/cocotb/osdtestlib/modtestclasses.py [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+        # wait for the active clock edge
+        yield RisingEdge(dut.clk)
+        # wait for the module to be ready to accept a data word
+        while not dut.debug_in_ready.value:
+            yield RisingEdge(dut.clk)
+            dut.debug_in.valid <= 0
+
+            
+    @cocotb.coroutine    
+    def init(self, dut, ID):
```
Comment (imphil) [inline test/cocotb/osdtestlib/modtestclasses.py]: This class is only driving the NoC. The testbench setup shouldn't be part of that class, instead it goes into the code which calls this class.

File: test/cocotb/osdtestlib/modtestclasses.py [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+    
+        # set module address to the second function parameter
+        dut.id <= ID
+
+        for _ in range(2):
+            yield RisingEdge(dut.clk)
+        dut.rst <= 0
+    
+    
+class Reg_Access:
```
Comment (imphil) [inline test/cocotb/osdtestlib/modtestclasses.py]: camel-case please

File: test/cocotb/osdtestlib/modtestclasses.py [hunk 5]
```diff
@@ -0,0 +1,637 @@
...
+    """
+    Class responsible for exchanging messages between debug modules and reading
```
Comment (imphil) [inline test/cocotb/osdtestlib/modtestclasses.py]: Wording as before: "Exchange messages between ..."

File: test/cocotb/osdtestlib/modtestclasses.py [hunk 6]
```diff
@@ -0,0 +1,637 @@
...
+            packet_subtype: choose from one of the const values: read or write 
+                            requests of variable length as well as read/write 
+                            responses(success/error).
+            message:        a payload of variable length (up to 8 16 bit words)
+        Returns:
+            -
+        """
+        
+        # build the 3 header flits in advance
+        dest = packet_utility.constructheader1(destination = destination)
```
Comment (imphil) [inline test/cocotb/osdtestlib/modtestclasses.py]: - Why do you pass the packet_utility class instance to the method? You can instead instantiate it directly in this method.
- There's no need to have three method calls to populate the header. Instead, you can have one function construct_header which sets all required fields. Or even better, have a class DiPacket, which represents a single packet in the debug interconnect (DI). You can then set the individual fields in this class and have one method get_flits() for example, which returns all the flits in the packet with the header, etc. set.

File: test/cocotb/osdtestlib/modtestclasses.py [hunk 7]
```diff
@@ -0,0 +1,637 @@
...
+            source:         id of the module which is the source.
+            packet_type:    possible values are: PACKET_TYPE_REG,
+                            PACKET_TYPE_PLAIN and PACKET_TYPE_EVENT.
+            packet_subtype: choose from one of the const values: read or write 
+                            requests of variable length as well as read/write 
+                            responses(success/error).
+            message:        a payload of variable length (up to 8 16 bit words)
+        Returns:
+            -
+        """
```
Comment (imphil) [inline test/cocotb/osdtestlib/modtestclasses.py]: This method is called send_packet, not construct_packet. Please split the two steps into two methods (also see the notes below on a separate class representing a packet).

File: test/cocotb/osdtestlib/modtestclasses.py [hunk 8]
```diff
@@ -0,0 +1,637 @@
...
+                            from.
+            response:       buffer variable which the response of the dut will 
+                            be stored in
+            
+        Returns:    
+            -
+        
+        """
+        
+        # branch depending on the width of the accessed data word
```
Comment (imphil) [inline test/cocotb/osdtestlib/modtestclasses.py]: The if branches differ only in the subtype and the expected response type. Set these two as variables depending on the width, and put all other code in the common code path (outside the ifs).

File: test/cocotb/osdtestlib/modtestclasses.py [hunk 9]
```diff
@@ -0,0 +1,637 @@
...
+                            register
+            response:       buffer variable which the response of the dut will 
+                            be stored in
+            
+            Returns:
+            -
+            
+        """
+        
+        # depending on how many data words are to be written into the memory
```
Comment (imphil) [inline test/cocotb/osdtestlib/modtestclasses.py]: dito

File: test/cocotb/osdtestlib/modtestclasses.py [hunk 10]
```diff
@@ -0,0 +1,637 @@
...
+            dut._log.info("Write process was successful!")
+        # if it returned an error, output that error
+        elif ((response[2] & 0x3C00) >> 10) == \
+        self.PACKET_SUBTYPE_REQ_WRITE_REG_ERROR:
+            dut._log.info("An error occurred during the write process!")
+        # otherwise something else went wrong
+        else:
+            dut._log.info("An unidentified error occurred!")
+                
+class Utility:
```
Comment (imphil) [inline test/cocotb/osdtestlib/modtestclasses.py]: Please use a more descriptive name. What kind of utility functions does this class contain? (Also, most likely all methods of this class will go away with a new class representing a packet.)

File: test/cocotb/osdtestlib/modtestclasses.py [hunk 11]
```diff
@@ -0,0 +1,637 @@
...
+            
+            # clear the buffer for the next RegRead()
+            del response[:]
+            
+        for _ in range(1):
+            yield RisingEdge(dut.clk)
+            
+            
+    @cocotb.coroutine        
+    def ModuleSubTest_STM(self, dut, test_driver, test_access, test_id):
```
Comment (imphil) [inline test/cocotb/osdtestlib/modtestclasses.py]: That should be part of the STM test, not of the generic test library.

**Comment by imphil** [comment]
Thanks @FrBier for this PR. This code needs a a bit more work before it can be merged.

- Don't commit any files are not code. This includes all the VCS project files and compiled python files (`.pyc`)
- Don't add the `osd_scm.sv` file in the test directory. It's not part of the test, but of the code to be tested and is included from there.
- Follow the PEP 8 coding style conventions for Python (https://www.python.org/dev/peps/pep-0008/). I've pointed out some problems in the code itself for the first iteration. 
- Place each class in a separate `.py` file with a name resembling the class name.
- Please make sure your files have no trailing whitespace (configure your editor to remove them on saving).

File: test/cocotb/osdtestlib/.synopsys_dve_rebuild.tcl [hunk 12]
```diff
@@ -0,0 +1,9 @@
+# DVE Simulation Rebuild/Restart Options
```
Comment (imphil) [inline test/cocotb/osdtestlib/.synopsys_dve_rebuild.tcl]: This file shouldn't be part of the commit.

File: test/cocotb/osdtestlib/.restartSimSession.tcl.old [hunk 13]
```diff
@@ -0,0 +1,195 @@
+# Begin_DVE_Session_Save_Info
```
Comment (imphil) [inline test/cocotb/osdtestlib/.restartSimSession.tcl.old]: This file shouldn't be part of the commit.

**Comment by FrBier** [comment]
I´ve added the makefile to execute the test in the stm/test directory.
I also tried:

cd modules/stm/test
../../../test/cocotb/cocotb_testrunner.py

as you suggested in the most recent issue.
However the terminal then responds with:

No test manifests (*.manifest.yaml) found in src/hardware/modules/stm/test

**Comment by imphil** [comment]
The error message is telling the exactly the problem: there is no test_stm.manifest.yaml file in modules/stm/test. Take inspiration by the one in scm to create it.

**Comment by FrBier** [comment]
I´ve added the manifest to the stm/test directory and adjusted it so that it should work for this module.
I´ve chosen the parameters as follows:

SYSTEM_VENDOR_ID: 1
NUM_MOD: 0

Both parameters were returned by the module after reading the corresponding registers.

SYSTEM_DEVICE_ID: 1

This value is chosen during the initialization of the test bench.


Unfortunately the test still can´t be executed. 
I receive the same error message that i get when trying to start VCS in a different test directory, e.g. in scm/test via the "make SIM=vcs SIM_ARGS=-gui" command:

[error_message.txt](https://github.com/opensocdebug/hardware/files/1374509/error_message.txt)

**Comment by imphil** [comment]
I don't see that error in my testing here. Please make sure that you don't have any old temporary files and build output from VCS around in your test directory or your home directory. And double-check that you didn't modify the Makefiles (or other code) in cocotb itself. 

(There are, however, other errors in your code that prevent compilation, but you should see those after fixing this first problem.)

**Comment by imphil** [comment]
On the other problems, the first the manifest doesn't contain all required files. This patch helps:

~~~
diff --git a/modules/stm/test/test_stm.manifest.yaml b/modules/stm/test/test_stm.manifest.yaml
index 4ed7c8f..ec1c132 100644
--- a/modules/stm/test/test_stm.manifest.yaml
+++ b/modules/stm/test/test_stm.manifest.yaml
@@ -5,11 +5,16 @@ sources:
   - ../../../blocks/buffer/common/dii_buffer.sv
   - ../../../blocks/regaccess/common/osd_regaccess.sv
   - ../../../blocks/regaccess/common/osd_regaccess_demux.sv
+  - ../../../blocks/regaccess/common/osd_regaccess_layer.sv
   - ../../../interconnect/common/debug_ring_expand.sv
   - ../../../interconnect/common/ring_router_demux.sv
   - ../../../interconnect/common/ring_router.sv
   - ../../../interconnect/common/ring_router_mux.sv
   - ../../../interconnect/common/ring_router_mux_rr.sv
+  - ../../../blocks/timestamp/common/osd_timestamp.sv
+  - ../../../blocks/buffer/common/osd_fifo.sv
+  - ../../../blocks/tracepacket/common/osd_trace_packetization.sv
+  - ../../../blocks/tracesample/common/osd_tracesample.sv
   - ../common/osd_stm.sv
 
 toplevel: osd_stm
~~~

Its still not running, I'm looking into it.

**Comment by imphil** [comment]
It seems we have a solution for the PYTHONPATH issue as well, I'm looking into it and get back to you in the afternoon.

**Comment by imphil** [comment]
I've now merged this into cocotb-tests and updated the test runner to find the python modules.

~~~
no56hud@philipp:/data/home/no56hud/src/opensocdebug-hardware/modules/stm/test$ ../../../test/cocotb/cocotb_testrunner.py 
make results.xml
make[1]: Entering directory '/data/home/no56hud/src/opensocdebug-hardware/modules/stm/test/objdir/osd_stm'
PYTHONPATH=/home/no56hud/src/cocotb/build/libs/x86_64:/home/no56hud/src/cocotb:/data/home/no56hud/src/opensocdebug-hardware/modules/stm/test:/data/home/no56hud/src/opensocdebug-hardware/test/cocotb:/data/home/no56hud/src/opensocdebug-hardware/test/cocotb:/data/home/no56hud/src/opensocdebug-hardware/modules/stm/test LD_LIBRARY_PATH=/home/no56hud/src/cocotb/build/libs/x86_64::/usr/lib:/usr/lib:/usr/lib:/usr/lib MODULE=test_stm TESTCASE= TOPLEVEL=osd_stm \
        TOPLEVEL_LANG= \
        sim_build/simv +define+COCOTB_SIM=1  -full64

Warning-[RT_UO] Unsupported option
  Unsupported option '-full64' is ignored

Chronologic VCS simulator copyright 1991-2017
Contains Synopsys proprietary information.
Compiler version M-2017.03-1_Full64; Runtime version M-2017.03-1_Full64;  Oct 11 15:03 2017
     -.--ns INFO     cocotb.gpi                                GpiCommon.cpp:91   in gpi_print_registered_impl       VPI registered
  0.00ns INFO     cocotb.gpi                                  gpi_embed.c:248  in embed_sim_init                  Running on Chronologic Simulation VCS Release  version M-2017.03-1_Full64
  0.00ns INFO     cocotb.gpi                                  gpi_embed.c:249  in embed_sim_init                  Python interpreter initialised and cocotb loaded!
  0.00ns INFO     cocotb                                      __init__.py:115  in _initialise_testbench           Running tests with Cocotb v1.0 from /home/no56hud/src/cocotb
  0.00ns INFO     cocotb                                      __init__.py:131  in _initialise_testbench           Seeding Python random module with 1507727033
  0.00ns INFO     cocotb.regression                         regression.py:167  in initialise                      Found test test_stm.test_stm
  0.00ns INFO     cocotb.regression                         regression.py:287  in execute                         Running test 1/1: test_stm
  0.00ns INFO     ..coroutine.test_stm.0x7fffe8c4f0d0       decorators.py:191  in send                            Starting test: "test_stm"
                                                                                                                  Description: 
                                                                                                                      Try accessing the design
                                                                                                                      
  0.00ns INFO     cocotb.osd_stm                              test_stm.py:26   in test_stm                        Resetting DUT
100.00ns INFO     cocotb.osd_stm                     ..ug_interconnect.py:217  in read_register                   Answer of the dut:
100.00ns INFO     cocotb.osd_stm                     ..ug_interconnect.py:218  in read_register                   [0, 1, 8192, 1]
100.00ns INFO     cocotb.osd_stm                     ..ug_interconnect.py:223  in read_register                   Read process was successful!
110.00ns INFO     cocotb.osd_stm                          eval_utility.py:56   in evaluate_response               Passed the test!
210.00ns INFO     cocotb.osd_stm                     ..ug_interconnect.py:217  in read_register                   Answer of the dut:
210.00ns INFO     cocotb.osd_stm                     ..ug_interconnect.py:218  in read_register                   [0, 1, 8192, 4]
210.00ns INFO     cocotb.osd_stm                     ..ug_interconnect.py:223  in read_register                   Read process was successful!
220.00ns INFO     cocotb.osd_stm                          eval_utility.py:56   in evaluate_response               Passed the test!
320.00ns INFO     cocotb.osd_stm                     ..ug_interconnect.py:217  in read_register                   Answer of the dut:
320.00ns INFO     cocotb.osd_stm                     ..ug_interconnect.py:218  in read_register                   [0, 1, 8192, 0]
320.00ns INFO     cocotb.osd_stm                     ..ug_interconnect.py:223  in read_register                   Read process was successful!
330.00ns INFO     cocotb.osd_stm                          eval_utility.py:56   in evaluate_response               Passed the test!
350.01ns INFO     cocotb.regression                         regression.py:223  in handle_result                   Test Passed: test_stm
350.01ns INFO     cocotb.regression                         regression.py:176  in tear_down                       Passed 1 tests (0 skipped)
350.01ns INFO     cocotb.regression                         regression.py:341  in _log_test_summary               ***************************************************************************
                                                                                                                  ** TEST               PASS/FAIL  SIM TIME(NS)  REAL TIME(S)  RATIO(NS/S) **
                                                                                                                  ***************************************************************************
                                                                                                                  ** test_stm.test_stm    PASS          350.01          0.03     11470.11  **
                                                                                                                  ***************************************************************************
                                                                                                                  
350.01ns INFO     cocotb.regression                         regression.py:358  in _log_sim_summary                *************************************************************************************
                                                                                                                  **                                 ERRORS : 0                                      **
                                                                                                                  *************************************************************************************
                                                                                                                  **                               SIM TIME : 350.01 NS                              **
                                                                                                                  **                              REAL TIME : 0.14 S                                 **
                                                                                                                  **                        SIM / REAL TIME : 2462.33 NS/S                           **
                                                                                                                  *************************************************************************************
                                                                                                                  
350.01ns INFO     cocotb.regression                         regression.py:184  in tear_down                       Shutting down...
           V C S   S i m u l a t i o n   R e p o r t 
Time: 350010 ps
CPU Time:      0.580 seconds;       Data structure size:   8.0Mb
Wed Oct 11 15:03:53 2017
/home/no56hud/src/cocotb/makefiles/simulators/Makefile.vcs:74: recipe for target 'results.xml' failed
make[1]: [results.xml] Error 4 (ignored)
make[1]: Leaving directory '/data/home/no56hud/src/opensocdebug-hardware/modules/stm/test/objdir/osd_stm'
~~~


Please see if you can reproduce this with the cocotb-test branch and then you can continue working on the STM stuff in the cocotb-test branch.
